src/Microsoft.AspNetCore.Http/StreamPipeWriter.cs

```diff
@@ -0,0 +1,320 @@
+        public override void Complete(Exception exception = null)
+        {
+            if (_isCompleted)
+            {
+                return;
+            }
+
+            _isCompleted = true;
+            if (exception != null)
+            {
+                _exceptionInfo = ExceptionDispatchInfo.Capture(exception);
+            }
+
+            _internalTokenSource?.Dispose();
+
+            if (_completedSegments != null)
+            {
+                foreach (var segment in _completedSegments)
+                {
+                    segment.Return();
+                }
+            }
+
+            _currentSegmentOwner?.Dispose();
+        }
+
+        /// <inheritdoc />
+        public override void OnReaderCompleted(Action<Exception, object> callback, object state)
+        {
+            throw new NotSupportedException("OnReaderCompleted isn't supported in StreamPipeWrapper.");
+        }
+
+        /// <inheritdoc />
+        public override ValueTask<FlushResult> FlushAsync(CancellationToken cancellationToken = default)
+        {
+            if (_bytesWritten == 0)
+            {
+                return new ValueTask<FlushResult>(new FlushResult(isCanceled: false, IsCompletedOrThrow()));
+            }
+
+            return FlushAsyncInternal(cancellationToken);
+        }
+
+        private void Cancel()
+        {
+            InternalTokenSource.Cancel();
+        }
+
+        private async ValueTask<FlushResult> FlushAsyncInternal(CancellationToken cancellationToken = default)
+        {
+            // Write all completed segments and whatever remains in the current segment
+            // and flush the result.
+            CancellationTokenRegistration reg = new CancellationTokenRegistration();
+            if (cancellationToken.CanBeCanceled)
+            {
+                reg = cancellationToken.Register(state => ((StreamPipeWriter)state).Cancel(), this);
+            }
+            using (reg)
+            {
+                var localToken = InternalTokenSource.Token;
+                try
+                {
+                    if (_completedSegments != null && _completedSegments.Count > 0)
+                    {
+                        var count = _completedSegments.Count;
+                        for (var i = 0; i < count; i++)
+                        {
+                            var segment = _completedSegments[0];
+#if NETCOREAPP2_2
+                            await _writingStream.WriteAsync(segment.Buffer.Slice(0, segment.Length), localToken);
+#elif NETSTANDARD2_0
+                            MemoryMarshal.TryGetArray<byte>(segment.Buffer, out var arraySegment);
+                            await _writingStream.WriteAsync(arraySegment.Array, 0, segment.Length, localToken);
+#else
+#error Target frameworks need to be updated.
+#endif
+                            _bytesWritten -= segment.Length;
+                            segment.Return();
+                            _completedSegments.RemoveAt(0);
```

Comment:
For next PR: this is slow and we don't have a benchmark that exercises this code path at all.